### ISSUE TITLE
Redundant return highlights too much

### ISSUE THREAD

**Issue by ndmitchell** [ndmitchell/hlint#1141]
Given:

```
main = do
    print 1
    print 2
    print 3
    return ()
    print 4
```

We get:

```
Sample.hs:(2,8)-(7,11): Warning: Redundant return
Found:
  do print 1
     print 2
     print 3
     return ()
     print 4
Perhaps:
  do print 1
     print 2
     print 3
     print 4

1 hint
```

Given our direction of travel, it would be much simpler just to highlight the return you should remove, rather than the whole do block. Much nicer in an IDE. Much easier to see what is wrong. Harder to see why it's redundant, but in an IDE, that's not a big deal.

Based on feedback from https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4089/diffs#note_300625

**Comment by Kleidukos** [comment]
I wholeheartedly support this! Thank you for opening this Neil! :)

**Comment by rmehri01** [comment]
Hi, I'd like to give this a try if possible!

**Comment by ndmitchell** [comment]
@rmehri01 great! Please do. Easiest thing is take the above sample, confirm it gives a range that is too big, then reduce the range.

As to tips:

* I recommend reading the contributing hints first: https://github.com/ndmitchell/hlint#hacking-hlint
* The code is very probably in https://github.com/ndmitchell/hlint/blob/7fad7d1344faed1b2f77a0caa4c6401ebb50a0ff/src/Hint/Monad.hs. I'd change the phrase "Redundant" in that module to "Redundant1" through "Redundant8" or whatever, so you can see which line is creating the hint, then fiddle around with locations after that. Most likely you replace the first expression argument with just the return line, and the second with Nothing, and see what you get.

**Comment by rmehri01** [comment]
@ndmitchell Awesome, thank you! :)

**Comment by rmehri01** [comment]
So I tried just replacing the second expression with an empty string but that prints `[]` since its put through `unsafePrettyPrint`. Instead I changed it to use `ideaRemove` and it seems to work but I'm not sure if it is the best solution. 

This is what I have right now:

```haskell
-- Rewrite 'do return x; $2' as 'do $2'.
monadStep wrap os@(o@(L _ (BodyStmt _ (fromRet -> Just (ret, _)) _ _ )) : xs@(_:_))
  = [ideaRemove Hint.Type.Warning ("Redundant " ++ ret) (getLoc o) (unsafePrettyPrint o) [Delete Stmt (toSS o)]]
```

Which produces the result:

```
Warning: Redundant return
Found:
  return ()
Perhaps you should remove it.
```

What do you think?

**Comment by ndmitchell** [comment]
Looks good to me. Want to raise a PR?
